sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridPublisherClient.cs

```diff
@@ -164,38 +172,54 @@ public virtual Response PublishCloudEvents(IEnumerable<CloudEvent> events, Cance
+            using DiagnosticScope scope = _clientDiagnostics.CreateScope($"{nameof(EventGridPublisherClient)}.{nameof(PublishCloudEvents)}");
             scope.Start();
 
             try
             {
-                List<CloudEvent> eventsWithSerializedPayloads = new List<CloudEvent>();
+                // List of events cannot be null
+                Argument.AssertNotNull(events, nameof(events));
+
+                List<CloudEventInternal> eventsWithSerializedPayloads = new List<CloudEventInternal>();
                 foreach (CloudEvent cloudEvent in events)
                 {
-                    // The 'Data' field is optional for the CloudEvent spec
-                    if (cloudEvent.Data != null)
+                    // Individual events cannot be null
+                    Argument.AssertNotNull(cloudEvent, nameof(cloudEvent));
+
+                    CloudEventInternal newCloudEvent = new CloudEventInternal(
+                        cloudEvent.Id,
+                        cloudEvent.Source,
+                        cloudEvent.Type,
+                        cloudEvent.SpecVersion)
                     {
-                        eventsWithSerializedPayloads.Add(
-                            new CloudEvent(
-                                cloudEvent.Id,
-                                cloudEvent.Source,
-                                cloudEvent.Type,
-                                cloudEvent.Specversion)
-                            {
-                                Data = new EventGridSerializer(
-                                    cloudEvent.Data,
-                                    _serializer,
-                                    cancellationToken),
-                                Time = cloudEvent.Time,
-                                Datacontenttype = cloudEvent.Datacontenttype,
-                                Dataschema = cloudEvent.Dataschema,
-                                Subject = cloudEvent.Subject
-                            });
+                        Time = cloudEvent.Time,
+                        Datacontenttype = cloudEvent.DataContentType,
+                        Dataschema = cloudEvent.DataSchema,
+                        Subject = cloudEvent.Subject
+                    };
+
+                    foreach (string key in cloudEvent.ExtensionAttributes.Keys)
+                    {
+                        cloudEvent.ExtensionAttributes.TryGetValue(key, out object value);
+                        newCloudEvent.Add(key, new EventGridSerializer(value, _serializer, cancellationToken));
                     }
-                    else
+
+                    // The 'Data' property is optional for CloudEvents
+                    if (cloudEvent.Data != null)
                     {
-                        eventsWithSerializedPayloads.Add(cloudEvent);
+                        if (cloudEvent.Data.GetType() == typeof(byte[]))
```

Comment:
I think ReadOnlySpan is a struct so I wasn't able to set that for `data` - is there still a way to test this?

Reply:
Ah, this is because it is a ref struct, and has other [limitations](https://docs.microsoft.com/en-us/dotnet/api/system.readonlyspan-1?view=netcore-3.1#remarks)
I think for now we can leave off ReadOnlySpan.